src/V3Sched.cpp

```diff
@@ -0,0 +1,964 @@
+        FileLine* const flp = topScopep->fileline();
+        AstVarRef* const vrefp = new AstVarRef{flp, m_vscp, VAccess::READ};
+        AstCMethodHard* callp = new AstCMethodHard{flp, vrefp, "at", new AstConst{flp, 0}};
+        callp->dtypeSetBit();
+        callp->pure(true);
+        AstSenItem* const senItemp = new AstSenItem{flp, VEdgeType::ET_TRUE, callp};
+        AstSenTree* const resultp = new AstSenTree{flp, senItemp};
+        topScopep->addSenTreep(resultp);
+        return resultp;
+    }
+
+    // Utility that assigns the 0 index trigger to fire when the given variable is zero
+    void addFirstIterationTriggerAssignment(AstVarScope* counterp) const {
+        FileLine* const flp = counterp->fileline();
+        AstVarRef* const vrefp = new AstVarRef{flp, m_vscp, VAccess::WRITE};
+        AstCMethodHard* callp = new AstCMethodHard{flp, vrefp, "at", new AstConst{flp, 0}};
+        callp->dtypeSetBit();
+        callp->pure(true);
+        m_funcp->stmtsp()->addHereThisAsNext(new AstAssign{
+            flp, callp,
+            new AstEq{flp, new AstVarRef{flp, counterp, VAccess::READ}, new AstConst{flp, 0}}});
+    }
+};
+
+const TriggerKit createTriggers(AstNetlist* netlistp, SenExprBuilder& senExprBuilder,
+                                std::vector<const AstSenTree*> senTreeps, const string& name,
+                                unsigned extra, bool slow = false) {
+    // FIXMEV5: don't generate empty trigger vectors
+
+    AstTopScope* const topScopep = netlistp->topScopep();
+    AstScope* const scopeTopp = topScopep->scopep();
+    FileLine* const flp = scopeTopp->fileline();
+
+    std::unordered_map<const AstSenTree*, AstSenTree*> map;
+
+    const uint32_t nTriggers = senTreeps.size() + extra;
+
+    // Create the TRIGGERVEC variable
+    AstBasicDType* const tDtypep = new AstBasicDType(flp, VBasicDTypeKwd::TRIGGERVEC,
+                                                     VSigning::UNSIGNED, nTriggers, nTriggers);
+    netlistp->typeTablep()->addTypesp(tDtypep);
+    AstVarScope* const vscp = scopeTopp->createTemp("__V" + name + "Triggered", tDtypep);
+
+    // Create the trigger computation function
+    AstCFunc* const funcp = makeSubFunction(netlistp, "_eval_triggers__" + name, slow);
+
+    // Create the trigger dump function (for debugging)
+    AstCFunc* const dumpp = makeSubFunction(netlistp, "_dump_triggers__" + name, slow);
+    dumpp->ifdef("VL_DEBUG");
+
+    // Add a print to the dumping function if there are no triggers pending
+    {
+        AstCMethodHard* const callp
+            = new AstCMethodHard{flp, new AstVarRef{flp, vscp, VAccess::READ}, "any"};
+        callp->dtypeSetBit();
+        AstIf* const ifp = new AstIf{flp, callp};
+        dumpp->addStmtsp(ifp);
+        ifp->addElsesp(
+            new AstText{flp, "VL_DBG_MSGF(\"         No triggers active\\n\");\n", true});
+    }
+
+    // Create a reference to a trigger flag
+    const auto getTrigRef = [&](uint32_t index, VAccess access) {
```

Comment:
I'd be inclined to make this and next lambda into functions, as the enclosing function is very large.

Reply:
As discussed, these capture some locals, so would be a bit tedious, so leaving them for now.